Remove useref and Fleet update changes

SahilArora92 opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [ ] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [x] Other, please clarify:
    -  Removing @useref usage from typespec files as that is not supported in sdk generation. Moved the swagger from Compute to Fleet namespace in `baseVirtualMachineProfile.tsp` and `common.tsp`. This required me to convert compute swagger to typespec as well.
    - Suppressed casing and other errors as swagger cannot be changed from Compute. 

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories")
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

Commit:
Remove useref and Fleet update changes

MaryGao commented on specification/azurefleet/AzureFleet.Management/common.tsp:
```diff
@@ -1,19 +1,44 @@
 import "@typespec/http";
 import "@typespec/rest";
 import "@typespec/versioning";
-import "@azure-tools/typespec-autorest";
```
Goo to see we removed this dependency!

msyyc commented:
https://github.com/Azure/sdk-release-request/issues/5257 is waiting for the PR

MaryGao commented on specification/azurefleet/AzureFleet.Management/common.tsp:
```diff
@@ -1,19 +1,44 @@
+  exceptionType?: string;
+
+  @doc("The internal error message or exception dump.")
+  errorDetail?: string;
```
According to definition it should be `errordetail` and `exceptiontype`, https://github.com/MaryGao/azure-rest-api-specs/blob/51d8dc5c05cdaeb436b41dc9ae7336ebf7b1e121/specification/compute/resource-manager/models.tsp#L3298-L3308.

SahilArora92 replied:
In the server definition it is camel case: https://msazure.visualstudio.com/One/_git/Compute-CPlat-Core?path=/src/Shared/Lib/ResourceProvider.Service/API/RpcEntityModel.cs&version=GBmaster&line=128&lineEnd=144&lineStartColumn=9&lineEndColumn=38&lineStyle=plain&_a=contents

SahilArora92 replied:
cc @grizzlytheodore to confirm if this was missed in the swagger.

SahilArora92 replied:
Discussed offline and the casing should be camel case.

XiaofeiCao replied:
@SahilArora92 In current Compute Swagger, it's not camel case, may I know the reason?
https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/common-types/v1/common.json#L59-L66

SahilArora92 replied:
They have it set up wrong. I confirmed from the compute spec release manager. 
![image](https://github.com/Azure/azure-rest-api-specs/assets/7030203/5dfc0c63-5e75-4584-88c3-dd8e522ea479)

XiaofeiCao commented on specification/azurefleet/AzureFleet.Management/common.tsp:
```diff
@@ -1,19 +1,44 @@
 @doc("ApiError for Fleet")
-@useRef("../../../../../compute/resource-manager/Microsoft.Compute/common-types/v1/common.json#/definitions/ApiError")
-model ApiError {}
+model ApiError is ApiErrorBase {
```
There is error type `ErrorDetail` in common-types, please use that:
```typescript
model ApiError {
  ...ErrorDetail,
  innererror?: InnerError,
}

@doc("Inner error details.")
model InnerError {
  @doc("The exception type.")
  exceptiontype?: string;

  @doc("The internal error message or exception dump.")
  errordetail?: string;
}
```

SahilArora92 replied:
Our Api does not use this errorDetail. It uses what [Microsoft.Compute](https://msazure.visualstudio.com/One/_git/Compute-CPlat-Core?path=/src/Shared/Lib/ResourceProvider.Service/API/RpcEntityModel.cs&version=GBmaster&line=123&lineEnd=125&lineStartColumn=1&lineEndColumn=43&lineStyle=plain&_a=contents) uses. Reference in [Fleet](https://msazure.visualstudio.com/One/_git/Compute-CPlat-Fleet?path=/src/Dev/FleetManager.ApiEntityModel/ApiEntityModel-FleetVMScaleSets.cs&version=GCec3796d65d47fc5b0e4170ea49fc2a4d51e768e9&line=51&lineEnd=51&lineStartColumn=1&lineEndColumn=31&lineStyle=plain&_a=contents)

Commit:
Remove useref

SahilArora92 commented on specification/azurefleet/AzureFleet.Management/baseVirtualMachineProfile.tsp:
```diff
@@ -0,0 +1,2213 @@
+  /**
+   * Json formatted public settings for the extension.
+   */
+  settings?: Record<unknown>;
```
Can I use this and suppress its warnings? or need to replace it with something else?

MaryGao replied:
@tadelesh I remember we decided on how to express swagger `type:object` in typespec? Could you help there?
https://github.com/Azure/azure-rest-api-specs/blob/d715112415eb08572ee96cc4bd77217aa44dedf1/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2024-03-01/virtualMachineScaleSet.json#L5002C1-L5009C11

SahilArora92 replied:
Yes this will always be a dictionary. 

I checked in Compute and they are defined as empty objects:
[azure-rest-api-specs/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2024-03-01/virtualMachineScaleSet.json at d715112415eb08572ee96cc4bd77217aa44dedf1 · Azure/azure-rest-api-specs (github.com)](https://github.com/Azure/azure-rest-api-specs/blob/d715112415eb08572ee96cc4bd77217aa44dedf1/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2024-03-01/virtualMachineScaleSet.json#L5002C1-L5009C11)

SahilArora92 replied:
Is there a way to define object in typespec but with no additionalProperties added in generated swagger?  @tadelesh 
![image](https://github.com/Azure/azure-rest-api-specs/assets/7030203/7cb366c2-3688-4c24-97c1-e8db5e495dbc)

Commit:
fix spell check

Commit:
Fix validation errors

Commit:
suppress casing errors

Commit:
Fix formatting

Commit:
Refactor

Commit:
Remove Fleet Forbidden values

Commit:
Add additional property suppression.

Commit:
edit suppressions

Commit:
Add Suppression for camel casing

SahilArora92 commented:
Please help in merging the PR as I have addressed all the comments. @MaryGao @tadelesh @XiaofeiCao

MaryGao commented on specification/azurefleet/AzureFleet.Management/examples/2024-05-01-preview/Fleets_Get.json:
```diff
@@ -314,11 +314,6 @@
                 "extensionsTimeBudget": "mbhjahtdygwgyszdwjtvlvtgchdwil"
               },
               "licenseType": "v",
-              "priority": "Regular",
```
why we have this change?

SahilArora92 replied:
These are forbidden fields which we are blocking for customers. Hence I removed them. Earlier there was no way to hide them as we were using useref. But now we don't want them as we define priority on our side and don't want customers to be confused.

TimLovellSmith replied:
Did you do breaking changes review of this?

ramoka178 commented:
>     "version": "2023-11-01-preview",

same comments as in this file for api-version 2024-05-01-preview too.

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:5 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>       "description": "Specifies additional XML formatted information that can be included in the\nUnattend.xml file, which is used by Windows Setup. Contents are defined by\nsetting name, component name, and the pass in which the content is applied.",

Here and at other places , are the "\n" needed?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:491 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "type": "string",

Consider adding "format": arm-id" if applicable , here and every place where resource-id is mentioned

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:528 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>         "licenseType": {

can this be made enum ?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:682 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "description": "Specifies the operating system settings for the virtual machines in the scale\nset."

needed? here and at other places

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:660 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "description": "Specifies the boot diagnostic settings state. Minimum api-version: 2015-06-15."

what does Minimum api-version represent, here and at other places ? 

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:676 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "description": "Specifies the time in which this VM profile for the Virtual Machine Scale Set\nwas created. Minimum API version for this property is 2023-09-01. This value\nwill be added to VMSS Flex VM tags when creating/updating the VMSS VM Profile\nwith minimum api-version 2023-09-01.",

consider adding timezone and format ?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:717 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>         "enabled": {

Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible. 

A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated]. 

Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its cofusing, and no better than a boolean. 

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:726 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "$ref": "#/definitions/FleetProperties",

this has "ProvisioningState" as a new property. so, now this is sent in the response for same api-version. If there are existing users for this api-version, will that not break customers ?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1166 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "description": "Resource Id"

Is this an azure resource id ? 
Can use "format": arm-id" ?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:985 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>         "userAssignedIdentityResourceId": {

"format": "arm-id" ?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1029 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "description": "Resource Id"

add more details on what type of id ? 
"format": "arm-id" 

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1201 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>         "provisionVMAgent": {

enum instead of bool, as mentioned at other place? 

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1305 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "type": "boolean",

enum instead of boolean? 

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1576 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "description": "Length of time a Virtual Machine being reimaged or having its OS upgraded will\nhave to potentially approve the OS Image Scheduled Event before the event is\nauto approved (timed out). The configuration is specified in ISO 8601 format,\nand the value must be 15 minutes (PT15M)"

curious, should it always be 15 mins ? or PT15M an example ?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1573 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>     "VirtualMachineScaleSetExtensionProperties": {

most of them are matching VirtualMachineExtensionProperties. Can such cases be consolidated into a definition and be used where needed ? avoids duplicate code.

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:2622 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>         },

id,name,type combo looks to be used at many places. Can they be consolidated into a definition and be used ?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:2598 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

ramoka178 commented:
>           "$ref": "#/definitions/FleetProperties",

Also, Could you update examples reflecting this change ? 

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1166 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

weidongxu-microsoft commented on specification/azurefleet/AzureFleet.Management/baseVirtualMachineProfile.tsp:
```diff
@@ -0,0 +1,2515 @@
+   * Windows](https://docs.microsoft.com/azure/virtual-machines/extensions/key-vault-windows).
+   */
+  #suppress "@azure-tools/typespec-azure-resource-manager/missing-x-ms-identifiers" "id exists inside sourceVault"
+  @OpenAPI.extension("x-ms-identifiers", ["sourceVault/id"])
```
@ArcturusZhang 
Does .NET have typespec corresponding to `x-ms-identifiers`?

weidongxu-microsoft commented on specification/azurefleet/AzureFleet.Management/baseVirtualMachineProfile.tsp:
```diff
@@ -0,0 +1,2515 @@
+    repair disks on Azure Linux VMs using the VMAccess
+    Extension](https://docs.microsoft.com/troubleshoot/azure/virtual-machines/troubleshoot-ssh-connection)
+    """)
+  adminPassword?: string;
```
They are `x-ms-secret` in Swagger https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2024-03-01/computeRPCommon.json#L1502

Should it be `@visibility("create", "update")`?

SahilArora92 replied:
Good point. Addressed this.

SahilArora92 commented:
> > ```
> > "version": "2023-11-01-preview",
> > ```
> 
> same comments as in this file for api-version 2024-05-01-preview too.
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:5 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)

What does this comment mean? Are you asking me to define the suppressions again in 2023-11-01-preview version? Something like this: https://github.com/Azure/azure-rest-api-specs/blob/1c63635d66ae38cff18045ab416a6572d3e15f6e/specification/playwrighttesting/resource-manager/readme.md?plain=1#L42

Commit:
Make confidential fields to have create and update visibility only.

Commit:
Merge branch 'main' into update-azurefleet-useref-and-scvmm

SahilArora92 commented:
> > ```
> >       "$ref": "#/definitions/FleetProperties",
> > ```
> 
> Also, Could you update examples reflecting this change ?
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1166 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)

I did run oav validate examples and there are no errors or issues with the change.

SahilArora92 commented:
> > ```
> >     },
> > ```
> 
> id,name,type combo looks to be used at many places. Can they be consolidated into a definition and be used ?
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:2598 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)

These definitions were generated using https://azure.github.io/typespec-azure/docs/howtos/migrate-swagger/get-started. So we can keep it as is.

SahilArora92 commented:
> "format": "arm-id"

This is using a sub resource in Compute swagger. Hence keeping it the same. https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/common-types/v1/common.json#L149

ramoka178 commented:
Not all comments have been addressed. Please address them and place it back in the queue.

SahilArora92 commented:
> > ```
> > "VirtualMachineScaleSetExtensionProperties": {
> > ```
> 
> most of them are matching VirtualMachineExtensionProperties. Can such cases be consolidated into a definition and be used where needed ? avoids duplicate code.
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:2622 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)

Most of these properties are a lift and shift from [compute swagger](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2024-03-01/virtualMachineScaleSet.json#L4976C6-L4976C47) as direct referencing is not allowed. Hence keeping it the same for consistency.

Commit:
update docs

SahilArora92 commented:
> curious, should it always be 15 mins ? or PT15M an example ?

The value should not exceed 15 min. Updated the docs.

SahilArora92 commented:
> > ```
> >     "enabled": {
> > ```
> 
> Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible.
> 
> A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated].
> 
> Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its cofusing, and no better than a boolean.
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:726 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)


Keeping it as boolean as it is boolean in Compute [swagger ](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2024-03-01/computeRPCommon.json#L1369)as well.

SahilArora92 commented:
> > ```
> >   "description": "Specifies additional XML formatted information that can be included in the\nUnattend.xml file, which is used by Windows Setup. Contents are defined by\nsetting name, component name, and the pass in which the content is applied.",
> > ```
> 
> Here and at other places , are the "\n" needed?
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:491 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)

\n is being added by typespec compiler automatically. I have not defined it in [typespec ](https://github.com/Azure/azure-rest-api-specs/pull/29493/files#diff-95b1ffba0937a1b59ac530c91aeada6a767c56bb1d971aee1070a0a3a7cd6004R1515) file.

MaryGao commented:
Regarding the [nextLink here](https://github.com/Azure/azure-rest-api-specs/blob/87a08b955c257c773a3bd42553c718d4b1092955/specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp#L56C3-L59C21), we have a `url` type in typespec, our recommandation is to use the native type if existing not format decorator? Could you update as following?

```
  #suppress "@azure-tools/typespec-azure-core/no-format" "NextLink is of uri format and need to be defined."
  @doc("The link to the next page of items")
  nextLink?: url;
```

Commit:
remove format decorator and use url type

Commit:
Merge branch 'main' into update-azurefleet-useref-and-scvmm

SahilArora92 commented:
> Regarding the [nextLink here](https://github.com/Azure/azure-rest-api-specs/blob/87a08b955c257c773a3bd42553c718d4b1092955/specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp#L56C3-L59C21), we have a `url` type in typespec, our recommandation is to use the native type if existing not format decorator? Could you update as following?
> 
> ```
>   @doc("The link to the next page of items")
>   nextLink?: url;
> ```
> 
> /cc @ArcturusZhang @tadelesh

Fixed

XiaofeiCao commented:
I've made a PR to fix `service-dir`:
https://github.com/Azure/azure-rest-api-specs/pull/29679

 This should only affect SDKs.

msyyc committed:
Merge branch 'main' into update-azurefleet-useref-and-scvmm

TimLovellSmith commented on specification/azurefleet/AzureFleet.Management/common.tsp:
```diff
@@ -1,19 +1,44 @@
-model ApiError {}
+model ApiError is ApiErrorBase {
+  @extension("x-ms-identifiers", ["message", "target"])
+  @doc("The Api error details")
```
>Api [](http://example.com/codeflow?start=12&length=3)

nit please capitalize acronyms like API as such, in documentation

TimLovellSmith commented on specification/azurefleet/AzureFleet.Management/common.tsp:
```diff
@@ -1,19 +1,44 @@
+  @doc("The exception type.")
+  exceptionType?: string;
+
+  @doc("The internal error message or exception dump.")
```
>@doc("The internal error message or exception dump.") [](http://example.com/codeflow?start=2&length=53)

This description suggest we return internal error and exception details which would not be an information disclosure risk, sounds like not a good security practice

SahilArora92 replied:
We are using the same ApiError which [Microsoft.Compute](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/common-types/v1/common.json#L63) uses currently hence keeping it the same.

TimLovellSmith commented:
>         },

needs to generate with x-ms-secret, add typescript annotations?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2024-05-01-preview/azurefleet.json:3002 in c39e954. [](commit_id = c39e954ad92de7959d7d87d327f4e78c6e349531, deletion_comment = False)

TimLovellSmith commented:
>         "protectedSettings": {

needs to generate with x-ms-secret, add typescript annotations?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:2469 in c39e954. [](commit_id = c39e954ad92de7959d7d87d327f4e78c6e349531, deletion_comment = False)

TimLovellSmith commented:
>         "protectedSettings": {

needs to generate with x-ms-secret, add typescript annotations?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:2659 in c39e954. [](commit_id = c39e954ad92de7959d7d87d327f4e78c6e349531, deletion_comment = False)

TimLovellSmith commented:
>         "adminPassword": {

needs to generate with x-ms-secret, add typescript annotations?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:2996 in c39e954. [](commit_id = c39e954ad92de7959d7d87d327f4e78c6e349531, deletion_comment = False)

TimLovellSmith commented:
>         "customData": {

Is the intention that this should be x-ms-secret annotated also, or is it not treated as a secret?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:3004 in c39e954. [](commit_id = c39e954ad92de7959d7d87d327f4e78c6e349531, deletion_comment = False)

TimLovellSmith commented:
>         "content": {

Is the intention that this should be x-ms-secret annotated also, or is it not treated as a secret?

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:517 in c39e954. [](commit_id = c39e954ad92de7959d7d87d327f4e78c6e349531, deletion_comment = False)

TimLovellSmith commented:
>         "protectedSettings": {

Looks like `@secret` is a thing...

---
In reply to: [2211459704](https://github.com/Azure/azure-rest-api-specs/pull/29493#issuecomment-2211459704) [](http://example.com/codeflow?ancestors=2211459704)

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:2659 in c39e954. [](commit_id = c39e954ad92de7959d7d87d327f4e78c6e349531, deletion_comment = False)

TimLovellSmith commented on specification/azurefleet/AzureFleet.Management/baseVirtualMachineProfile.tsp:
```diff
@@ -0,0 +1,2518 @@
+
+/**
+ * Describes the base virtual machine profile for fleet
+ */
```
user facing docs need to be `@doc` and not just doc comments, right?

SahilArora92 replied:
Is it documented any where? The [tool](https://azure.github.io/typespec-azure/docs/howtos/migrate-swagger/get-started) which converts swagger to typespec, generated the doc comments in this way.

SahilArora92 replied:
I made the changes in typespec to use `doc` instead of comments but there is no change in swagger. Hence keeping it the same. 
![image](https://github.com/Azure/azure-rest-api-specs/assets/7030203/45a6a3dc-e218-458a-b544-6d37edaf0c9f)
![image](https://github.com/Azure/azure-rest-api-specs/assets/7030203/c493e5cf-3d96-432b-9b15-b3c87f84be95)

TimLovellSmith commented on specification/azurefleet/AzureFleet.Management/baseVirtualMachineProfile.tsp:
```diff
@@ -0,0 +1,2518 @@
+   */
+  @visibility("read")
+  // FIXME: (utcDateTime) Please double check that this is the correct type for your scenario.
+  timeCreated?: utcDateTime;
```
I see a FIXME warning here.
Please verify you can remove it

SahilArora92 replied:
Updated

TimLovellSmith commented on specification/azurefleet/AzureFleet.Management/baseVirtualMachineProfile.tsp:
```diff
@@ -0,0 +1,2518 @@
+  /**
+   * The time of the status.
+   */
+  // FIXME: (utcDateTime) Please double check that this is the correct type for your scenario.
```
>FIXME [](http://example.com/codeflow?start=5&length=5)

I see a FIXME warning here.
Please verify you can remove it

SahilArora92 replied:
updated

SahilArora92 commented:
> protectedSettings

I cannot add @secret to Record unknown. 
![image](https://github.com/Azure/azure-rest-api-specs/assets/7030203/4df34e8e-f9f5-4252-9878-1e0e5732e45b)

Commit:
Address comments

SahilArora92 commented:
> > ```
> >     "content": {
> > ```
> 
> Is the intention that this should be x-ms-secret annotated also, or is it not treated as a secret?
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:517 in [c39e954](https://github.com/Azure/azure-rest-api-specs/commit/c39e954ad92de7959d7d87d327f4e78c6e349531). [](commit_id = [c39e954](https://github.com/Azure/azure-rest-api-specs/commit/c39e954ad92de7959d7d87d327f4e78c6e349531), deletion_comment = False)

made the necessary changes

Commit:
revert settings change as it is visible.

Commit:
Address comments

SahilArora92 commented:
> > > protectedSettings
> > 
> > 
> > I cannot add @secret to Record unknown. ![image](https://private-user-images.githubusercontent.com/7030203/346456876-4df34e8e-f9f5-4252-9878-1e0e5732e45b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1MTA3MzcsIm5iZiI6MTcyMDUxMDQzNywicGF0aCI6Ii83MDMwMjAzLzM0NjQ1Njg3Ni00ZGYzNGU4ZS1mOWY1LTQyNTItOTg3OC0xZTBlNTczMmU0NWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDlUMDczMzU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDY2MzQ2YzQzMjE2NzhkOGRkNDk0ZTVkOTMzMTA0NjI3NDNjM2M1MTY0OWY5ZWM0OTAyMTUyZGY5OGM2MjMxZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.kKVuxaQUrYiDbYxStXxnKM5oCfN8c2Dq22yvprnKcic)
> 
> `@extension("x-ms-secret", true)` may work.
> 
> This is not great typespec, but it may get the Swagger correct... (typespec seems mostly on [`@visibility`](https://azure.github.io/typespec-azure/docs/next/howtos/ARM/resource-type#property-visibility-and-other-constraints). Not an expert though.)

Fixed by using @OpenAPI.extension()

Commit:
configReference is visible in response.

madhura-np commented:
There are still some comments unaddressed regarding booleans, armid, x-ms-secret annocation. please address/reply.

SahilArora92 commented:
> > ```
> >     },
> > ```
> 
> needs to generate with x-ms-secret also, add Typespec annotations?
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2024-05-01-preview/azurefleet.json:3002 in [c39e954](https://github.com/Azure/azure-rest-api-specs/commit/c39e954ad92de7959d7d87d327f4e78c6e349531). [](commit_id = [c39e954](https://github.com/Azure/azure-rest-api-specs/commit/c39e954ad92de7959d7d87d327f4e78c6e349531), deletion_comment = False)

Added

SahilArora92 commented:
> > ```
> >       "type": "boolean",
> > ```
> 
> enum instead of boolean?
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1576 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)

It cannot be changed as it is defined in CRP and they are defined as boolean in ComputeRP

Commit:
Add arm-id format to resource ids

SahilArora92 commented:
> There are still some comments unaddressed regarding booleans, armid, x-ms-secret annocation. please address/reply.

Addressed those comments. And added arm-id formatting and x-ms-secret.

madhura-np commented:
>           "description": "Specifies the boot diagnostic settings state. Minimum api-version: 2015-06-15."

This would be good to address. 

---
In reply to: [2187445800](https://github.com/Azure/azure-rest-api-specs/pull/29493#issuecomment-2187445800) [](http://example.com/codeflow?ancestors=2187445800)

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:676 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

madhura-np commented:
>           "description": "Specifies the time in which this VM profile for the Virtual Machine Scale Set\nwas created. Minimum API version for this property is 2023-09-01. This value\nwill be added to VMSS Flex VM tags when creating/updating the VMSS VM Profile\nwith minimum api-version 2023-09-01.",

Can the timezone and format be added to description?

---
In reply to: [2187445876](https://github.com/Azure/azure-rest-api-specs/pull/29493#issuecomment-2187445876) [](http://example.com/codeflow?ancestors=2187445876)

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:717 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

madhura-np commented:
>           "$ref": "#/definitions/FleetProperties",

this is worth addressing/replying. 

---
In reply to: [2187446032](https://github.com/Azure/azure-rest-api-specs/pull/29493#issuecomment-2187446032) [](http://example.com/codeflow?ancestors=2187446032)

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1166 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

Commit:
Address comments

madhura-np commented:
>           "$ref": "#/definitions/FleetProperties",

FleetUpdate example can be appended to include provisioningState.

---
In reply to: [2187447150](https://github.com/Azure/azure-rest-api-specs/pull/29493#issuecomment-2187447150) [](http://example.com/codeflow?ancestors=2187447150)

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1166 in f557502. [](commit_id = f55750241673c2fbc3e7fdec4058c1568dafb9cf, deletion_comment = False)

SahilArora92 commented:
> booleans

enum can't be used instead of booleans as I cannot change the types released in Compute swagger.

> > ```
> >       "description": "Specifies the boot diagnostic settings state. Minimum api-version: 2015-06-15."
> > ```
> 
> This would be good to address.
> 
> In reply to: [2187445800](https://github.com/Azure/azure-rest-api-specs/pull/29493#issuecomment-2187445800)
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:676 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)

fixed. Removed unnecessary minimum api version

SahilArora92 commented:
> > ```
> >       "description": "Specifies the time in which this VM profile for the Virtual Machine Scale Set\nwas created. Minimum API version for this property is 2023-09-01. This value\nwill be added to VMSS Flex VM tags when creating/updating the VMSS VM Profile\nwith minimum api-version 2023-09-01.",
> > ```
> 
> Can the timezone and format be added to description?
> 
> In reply to: [2187445876](https://github.com/Azure/azure-rest-api-specs/pull/29493#issuecomment-2187445876)
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:717 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)

Added Timezone format in description.

SahilArora92 commented:
> > ```
> >       "$ref": "#/definitions/FleetProperties",
> > ```
> 
> this is worth addressing/replying.
> 
> In reply to: [2187446032](https://github.com/Azure/azure-rest-api-specs/pull/29493#issuecomment-2187446032)
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1166 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)

This change was proposed by the sdk team which resulted in Update fields to be removed. [Update the update model for swagger in azure fleet by MaryGao · Pull Request #29492 · Azure/azure-rest-api-specs (github.com)](https://github.com/Azure/azure-rest-api-specs/pull/29492/files) If I remove the `@parametervisibility` from ArmCustomPacthAsync then Fleet PropertiesUpdate remain unchanged. 
![image](https://github.com/user-attachments/assets/8bd40b88-48d9-4c9e-946f-d11f89fd1b92)
@MaryGao Can you please let us know why this change is needed?

Commit:
update provisioningState in examples

MaryGao commented:
> > > ```
> > >       "$ref": "#/definitions/FleetProperties",
> > > ```
> > 
> > 
> > this is worth addressing/replying.
> > In reply to: [2187446032](https://github.com/Azure/azure-rest-api-specs/pull/29493#issuecomment-2187446032)
> > Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2023-11-01-preview/azurefleet.json:1166 in [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf). [](commit_id = [f557502](https://github.com/Azure/azure-rest-api-specs/commit/f55750241673c2fbc3e7fdec4058c1568dafb9cf), deletion_comment = False)
> 
> Confirmed with Mary that we need provisioningState property in update response. This was missed earlier. So the current changes act as bugfix.
> 
> cc @madhura-np

@madhura-np I clarified with service team this property is existing in response since begining, and I think adding optional is allowed in new API version.

Commit:
update examples

raosuhas commented:
>         },

why do you need this ? I dont see this definition actually being used anywhere. 

Also location must have 'x-ms-mutability':['read', 'create'] extension defined as flaged in the errors 

---
Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2024-05-01-preview/azurefleet.json:1860 in 9704381. [](commit_id = 97043813edfc7ed3f4343ece7b4d970bb7d6fcb2, deletion_comment = False)

raosuhas commented:
> > 
> 
> Yeah adding the decorator parametervisibility is because we don't want to introduce un-necessary Update model in SDK side. Actually this will not impact the REST API behavior. The property `provisioningState` with optional and readonly is always there from service end side.
> 
> @madhura-np I clarified with service team this property is existing in response since begining, and I think adding optional is allowed in new API version.

I think that is fine

raosuhas commented:
> Also location must have 'x-ms-mutability':['read', 'create'] extension defined as flaged in the errors

Looks like this is used to define the VirtualMachineExtension type. Can you please explain what are you trying to do here in terms of modelling 

I dont tsee this in the examples as well. Could you please add an example here as well so that users can see how this property is to be used ?

Commit:
Merge branch 'main' into update-azurefleet-useref-and-scvmm

SahilArora92 commented:
> > ```
> >     },
> > ```
> 
> why do you need this ? I dont see this definition actually being used anywhere.
> 
> Also location must have 'x-ms-mutability':['read', 'create'] extension defined as flaged in the errors
> 
> Refers to: specification/azurefleet/resource-manager/Microsoft.AzureFleet/preview/2024-05-01-preview/azurefleet.json:1860 in [9704381](https://github.com/Azure/azure-rest-api-specs/commit/97043813edfc7ed3f4343ece7b4d970bb7d6fcb2). [](commit_id = [9704381](https://github.com/Azure/azure-rest-api-specs/commit/97043813edfc7ed3f4343ece7b4d970bb7d6fcb2), deletion_comment = False)

It does not have any visibility set to read or create. I cannot refer properties directly from compute namespace.  https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/common-types/v1/common.json#L172

Let me know if I still need to edit it?

Commit:
Update location in examples

SahilArora92 commented:
> > Also location must have 'x-ms-mutability':['read', 'create'] extension defined as flaged in the errors
> 
> Looks like this is used to define the VirtualMachineExtension type. Can you please explain what are you trying to do here in terms of modelling
> 
> I dont tsee this in the examples as well. Could you please add an example here as well so that users can see how this property is to be used ?

Example is present: 

> > Also location must have 'x-ms-mutability':['read', 'create'] extension defined as flaged in the errors
> 
> Looks like this is used to define the VirtualMachineExtension type. Can you please explain what are you trying to do here in terms of modelling
> 
> I dont tsee this in the examples as well. Could you please add an example here as well so that users can see how this property is to be used ?

Updated examples. Location was having a random value. https://github.com/Azure/azure-rest-api-specs/blob/4ac7e1ff5ec6bb7bc19a24557548b2d54dfd5287/specification/azurefleet/AzureFleet.Management/examples/2024-05-01-preview/Fleets_CreateOrUpdate.json#L394

weidongxu-microsoft commented on specification/azurefleet/AzureFleet.Management/baseVirtualMachineProfile.tsp:
```diff
@@ -0,0 +1,2524 @@
+   * List of virtual machine extensions to exclude when applying the Security
+   * Posture.
+   */
+  excludeExtensions?: VirtualMachineExtension[];
```
Just a heads-up. It may affect your service.

Note that there is recent break in CRP on this property.
https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2024-03-01/virtualMachineScaleSet.json#L6345-L6351

I do now know why the break. Just found this during upgrade of compute SDK...

SahilArora92 replied:
Oh, so we should also update to the latest and change its type to string[]?

weidongxu-microsoft replied:
It may depend on the CRP api-version your service uses for this response (and what the response really is).

PS: here is CRP's response about why this break
https://github.com/Azure/azure-rest-api-specs/pull/28891#issuecomment-2140744373

SahilArora92 replied:
Thanks. We are using a range of CRP versions. Is there a way to define a different model in typespec when a particular [version ](https://github.com/Azure/azure-rest-api-specs/blob/12a9609361d21154a0fb141dd682ba3f2a437d7f/specification/azurefleet/AzureFleet.Management/fleet.tsp#L102) is passed?

SahilArora92 replied:
Updated

Commit:
Update example

raosuhas commented:
As discussed in office hours consider if you want to allow customers to update the location field in the payload with a RE-PUT. Please restart the review once you have discussed this internally

Commit:
Update security posture reference to the latest compute api version.

SahilArora92 commented:
> As discussed in office hours consider if you want to allow customers to update the location field in the payload with a RE-PUT. Please restart the review once you have discussed this internally

Updated the model to the latest securityPostureReference which does not have location. Removed unnecessary models.

Commit:
Merge branch 'main' into update-azurefleet-useref-and-scvmm

Commit:
Fix merge conflicts

Commit:
Update and add more examples

SahilArora92 commented:
/azp run

SahilArora92 commented:
/azp run

SahilArora92 commented:
/pr RequestMerge
